documentation/man/rrlog.1
.\" Text automatically generated by txt2man
.TH rrlog 1 "16 June 2018" "man" "FreeBSD Reference Manual"
.SH NAME
\fBrrlog – display contents of replay log.
\fB
.SH SYNOPSIS
.nf
.fam C
  \fBrrlog\fP [\fIoptions\fP] \fIpath-to-logfile\fP

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
The \fBrrlog\fP utility is part of the Castor record/replay system. It displays
.TP
.B
the contents of a log which has been recorded using the \fBrecord\fP(1) utility.
and ends with a summary.
.SH OPTIONS
.TP
.B
\fB-h\fP
Help - displays summary of \fIoptions\fP.
.TP
.B
\fB-d\fP
Dump - display raw log records, interpretation is described in                       the log format section below.
.TP
.B
\fB-t\fP
Truss - truss mode, just display system calls.
.TP
.B
\fB-a\fP
All -  display all events. (default)
.TP
.B
\fB-h\fP
Help - display this message.
.TP
.B
\fB-s\fP
Summary - show summary of total # of events by type and thread.
.SH TRUSS MODE FORMAT
The truss mode displays most (see ALL mode for exceptions) of the system call
events in the log in a more easily readable format, similar to truss. It is
interpreted as follows: System call fields shown with an '_' denote fields 
that castor does not record. If a leading object e.g. file descriptor of object
has been recorded this is displayed inline e.g. in getrlimit. Subsequent fields
that have been recorded are shown by their name e.g. buf, rpl, and if castor
can parse and display their contents are shown in brackets after the return
value (which follows the '=' sign).
.RE
.PP
EXAMPLE:
.PP
.nf
.fam C
    $> rrlog -t ./unit-tests/udp.rr
    readlink(_, buf, _) = -1 [errno: ENOENT]
    issetugid() = 0
    getrlimit(3,rlp) = 0
    getpid() = 84465
    fstat(1,sb) = 0   [sb: { mode=-rw-r--r-- ,inode=605802,size=184,blksize=4096 }]

.fam T
.fi
.SH ALL MODE FORMAT
Includes everything from TRUSS mode, plus all other events logged in the core
runtime (as opposed to by the system call recording facility which is generated
programmatically) the runtime, including thread locking events, inline
assembly, and a small number of system calls. Often truss mode is
all that's needed to understand a problem, but sometimes ALL mode is
useful and nicer to look at than RAW mode.
.SH RAW LOG FORMAT
The log is a sorted collection of entries with the following fields which
are displayed in order:
.PP
"Event #", "Thread #", "Event", "Object ID", "Value[0] - Value[4]"
.PP
These fields denote the following:
.PP
Event # - order event was recorded/will be replayed.
.PP
Thread # - threadid of thread which logged the event.
.PP
Event - name of event that generated the log entry: the name will either
be of a system call e.g. FORK, WAIT, GETPID. Argument data for a system
call: DATA, A thread library call: e.g. MUTEX_INIT, COND_INIT, SPIN_LOCK.
or a compiler primitive: ATOMIC_LOAD, INLINE_ASM. Or additional castor
meta-data e.g. LOCKED_EVENT
.PP
Object ID - stores the object a system call is operating on e.g. the file
descriptor # for \fBread\fP(2) or the socket descriptor # for \fBsendmsg\fP(2).
.PP
VALUE[0-4] - contents of this field is system call dependant. In general
VALUE[0] stores the result of a system call, VALUE[1] stores the value of
errno if result < 0. Value fields can also be used to store the side
effects of system calls e.g. that are small enough to be stored with the
event and don't require seperate data records.
.RE
.PP

.SH EXIT STATUS
If an error occurs while reading the log exits 1, otherwise, exits 0.
.SH COMPATIBILITY
Castor only targets FreeBSD Release 1.2 Stable and beyond.
.SH SEE ALSO
\fBreplay\fP(1), \fBrecord\fP(1), \fBrr_printf\fP(3), \fBrr_assert\fP(3)


tools/rrlog/rrlog.c

#include <assert.h>
#include <stddef.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <unistd.h>
#include <fcntl.h>

#include <castor/archconfig.h>
#include <castor/rrshared.h>
#include <castor/events.h>
#include <castor/file_format.h>

#include "events_pretty_printer_gen.h"

enum display_modes {ALL_MODE, TRUSS_MODE, DUMP_MODE};

static int logfd;

struct {
    uint32_t	evtid;
    const char* str;
    void (*pretty_print)(RRLogEntry entry);
    uint64_t	count;
} rreventTable[] =
{
#define RREVENT(_a, _b) { RREVENT_##_a, #_a, pretty_print_##_a, 0 },
    RREVENT_TABLE
#undef RREVENT
    { 0, 0 }
};

uint64_t eventsPerThread[RRLOG_MAX_THREADS];

int openLog(const char * logfile) {
    castor_magic_t magic;
    castor_version_t version;
    const int flags = O_RDWR;

    int logfd = open(logfile, flags, 0600);

    if (logfd < 0) {
	fprintf(stderr, "Could not open record/replay log '%s'", logfile);
	perror("open");
	exit(1);
    }

    read(logfd, &magic, sizeof(magic));
    read(logfd, &version, sizeof(version));
    if (magic != CASTOR_MAGIC_NUMBER) {
	fprintf(stderr, "OpenLog failed: %s does not appear to be a log file, invalid "\
			"magic number at file start.\n", logfile);
	exit(1);
    }
    if (version != CASTOR_VERSION_NUMBER) {
	fprintf(stderr, "OpenLog failed: %s recorded with version %d,"\
			"current version is %d.\n",
			logfile, version, CASTOR_VERSION_NUMBER);
	exit(1);
    }

    return logfd;
}

void initStats()
{
    for (int i = 0; i < RRLOG_MAX_THREADS; i++) {
	eventsPerThread[i] = 0;
    }
}

void updateStats(RRLogEntry entry) {
    rreventTable[entry.event].count++;
    eventsPerThread[entry.threadId] += 1;
}

void displayStats() {
    printf("\n%-16s      %s\n", "Event", "Count");
    for (int i = 0; rreventTable[i].str != 0; i++) {
	if (rreventTable[i].count != 0) {
	    printf("%-16s      %lu\n", rreventTable[i].str, rreventTable[i].count);
	}
    }

    printf("\n%-8s      %s\n", "Thread", "# Events");
    for (int i = 0; i < RRLOG_MAX_THREADS; i++) {
	if (eventsPerThread[i] > 0) {
	    printf("%-8d      %lu\n", i, eventsPerThread[i]);
	}
    }
}

int readEntry(RRLogEntry * entry){
    int result = read(logfd, (void *)entry, sizeof(RRLogEntry));
    if (result < 0) {
	perror("read");
	return 0;
    }
    if (result != sizeof(*entry)) {
	return 0;
    }
    return 1;
}

static inline void assertData(RRLogEntry * entry) {
    if (entry->event != RREVENT_DATA) {
	fprintf(stderr, "invalide entry: something went horribly wrong while reading the log.");
	exit(1);
    }
}


void readData(uint8_t * buf, size_t len) {
    assert(buf != NULL);
    uint64_t recs = len / RREVENT_DATA_LEN;
    uint64_t rlen = len % RREVENT_DATA_LEN;
    RRLogEntry entry;

    for (unsigned int i = 0; i < recs; i++) {
	readEntry(&entry);
	assertData(&entry);
	uint8_t *src = ((uint8_t *)&entry) + RREVENT_DATA_OFFSET;
	memcpy(buf, src, RREVENT_DATA_LEN);
	buf += RREVENT_DATA_LEN;
    }
    if (rlen) {
	readEntry(&entry);
	assertData(&entry);
	uint8_t *src = ((uint8_t *)&entry) + RREVENT_DATA_OFFSET;
	memcpy(buf, src, rlen);
    }
}

void dumpEntry(RRLogEntry entry)
{
    if (entry.event > RREVENTS_MAX) {
	fprintf(stderr, "invalid event number %u\n", entry.event);
	exit(1);
    }
    const char * event_name = rreventTable[entry.event].str;

    printf("%016ld  %08x  %-16s  %016lx  %016lx  %016lx  %016lx  %016lx  %016lx\n",
	    entry.eventId, entry.threadId, event_name, entry.objectId,
	    entry.value[0], entry.value[1], entry.value[2],
	    entry.value[3], entry.value[4]);
}

void prettyPrintEntry(RRLogEntry entry)
{
    if (entry.event > RREVENTS_MAX) {
	fprintf(stderr, "invalid event number %u\n", entry.event);
	exit(1);
    }

    rreventTable[entry.event].pretty_print(entry);
}

void prettyPrintSyscalls(RRLogEntry entry)
{
    if (entry.event > RREVENTS_MAX) {
	fprintf(stderr, "invalid event number %u\n", entry.event);
	exit(1);
    }
    if (GENERATED_SYSCALLS_CONTAINS(entry.event) ||
	    BUILTIN_SYSCALLS_CONTAINS(entry.event)) {
	rreventTable[entry.event].pretty_print(entry);
    }
}


void
usage()
{
    printf("rrlog [options] [logfile]\n");
    printf("  -d            Dump - display raw log records.\n");
    printf("  -t            Truss - truss mode, just display system calls.\n");
    printf("  -a            All -  display all events. (default)\n");
    printf("  -h            Help - display this message.\n");
    printf("  -s            Summary - show summary of total # of events by type and thread.\n");
}


int main(int argc, char * const argv[])
{
    int mode = ALL_MODE;
    char ch;
    int showStats = 0;

    while ((ch = getopt(argc, argv, "dtash:")) != -1) {
	switch (ch) {
	    case 'd':
		mode = DUMP_MODE;
		break;
	    case 't':
		mode = TRUSS_MODE;
		break;
	    case 'a':
		mode = ALL_MODE;
		break;

	    case 's':
		showStats = 1;
		break;
	    case 'h':
		usage();
		exit(0);
		break;
	    default:
		usage();
		exit(1);
	}
    }
    argc -= optind;
    argv += optind;

    if (argc == 0) {
	usage();
	exit(1);
    }

    logfd = openLog(argv[0]);
    initStats();

    RRLogEntry entry;
    switch (mode) {
	case ALL_MODE:
	    while (readEntry(&entry)) {
		updateStats(entry);
		prettyPrintEntry(entry);
	    }
	    break;
	case TRUSS_MODE:
	    while (readEntry(&entry)) {
		updateStats(entry);
		prettyPrintSyscalls(entry);
	    }
	    break;
	case DUMP_MODE:
	    printf("%-16s  %-8s  %-16s  %-16s  %-16s  %-16s  %-16s  %-16s  %-16s\n",
		    "Event #", "Thread #", "Event", "Object ID",
		    "Value[0]", "Value[1]", "Value[2]", "Value[3]", "Value[4]");
	    while (readEntry(&entry)) {
		updateStats(entry);
		dumpEntry(entry);
	    }
	    break;
    }

    if (showStats) {
	displayStats();
    }

    return 0;
}

